catalyst/contrib/nn/criterion/ntxent.py

```diff
@@ -0,0 +1,79 @@
+import torch
+from torch import nn
+from math import e
+
+class NTXentLoss(nn.Module):
+    """A Contrastive embedding loss.
+
+    It has been proposed in `A Simple Framework
+    for Contrastive Learning of Visual Representations`_.
+
+    Example:
+
+    .. code-block:: python
+
+        import torch
+        from torch.nn import functional as F
+        from catalyst.contrib.nn import NTXentLoss
+
+        embeddings_left = F.normalize(torch.rand(256, 64, requires_grad=True))
+        embeddings_right = F.normalize(torch.rand(256, 64, requires_grad=True))
+        criterion = NTXentLoss(tau = 0.1)
+        criterion(embeddings_left, embeddings_right)
+
+    .. _`A Simple Framework for Contrastive Learning of Visual Representations`:
+        https://arxiv.org/abs/2103.03230
+    """
+
+    def __init__(self, tau: float, reduction: str = "mean") -> None:
+        """
+
+        Args:
+            tau: temperature
+            reduction (string, optional): specifies the reduction to apply to the output:
+                ``"none"`` | ``"mean"`` | ``"sum"``.
+                ``"none"``: no reduction will be applied,
+                ``"mean"``: the sum of the output will be divided by the number of
+                positive pairs in the output,
+                ``"sum"``: the output will be summed.
+        """
+        super().__init__()
+        self.tau = tau
+        self.cosine_sim = nn.CosineSimilarity()
+        self.reduction = reduction
+
+    def forward(self, features1: torch.Tensor, features2: torch.Tensor) -> torch.Tensor:
+        """
+
+        Args:
+            features1: batch with samples features of shape
+                [bs; feature_len]
+            features2: batch with samples features of shape
+                [bs; feature_len]
+
+        Returns:
+            torch.Tensor: NTXent loss
+        """
+        assert (
+            features1.shape == features2.shape
+        ), f"Invalid shape of input features: {features1.shape} and {features2.shape}"
+        bs = features1.shape[0]
+
+        feature_matrix = torch.cat([features1, features2])
+        feature_matrix = torch.nn.functional.normalize(feature_matrix)
+        cosine_matrix = (2 - torch.cdist(feature_matrix, feature_matrix) ** 2) / 2
```

Comment:
Can you explain, why is it cosine matrix? mb you're right, but I just don't understand the math.
Probably, the easier way to compute cosine similarity matrix of normalised vectors is just `torch.mm(feature_matrix, feature_matrix.transpose(0, 1))`